### PR TITLE
Implement Window Function Spilling

### PR DESCRIPTION
**Pull request by sachdevs**
Original PR: https://github.com/prestosql/presto/pull/228/
Broken up into ORDER BY spilling (merged in https://github.com/prestodb/presto/pull/14836) and Window function spilling (this one)

Part of https://github.com/prestodb/presto/issues/14935

TODO
- Release notes
- Window function docs

### CODE REVIEW

**Comment by wenleix** [comment]
@sachdevs  : is it possible also to mention the PR in PrestoSQL that you backport from?

**Comment by sachdevs** [comment]
Updated - it's the same as the ORDER BY spilling PR on prestoSQL's end.

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 1]
```diff
@@ -264,6 +264,12 @@ public AggregatedMemoryContext aggregateUserMemoryContext()
         return new InternalAggregatedMemoryContext(operatorMemoryContext.aggregateUserMemoryContext(), memoryFuture, this::updatePeakMemoryReservations, false);
     }
 
+    // caller shouldn't close this context as it's managed by the OperatorContext
+    public AggregatedMemoryContext aggregateRevocableMemoryContext()
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:268]: This looks quite general, surprised not added before when adding aggregation spilling

File: presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java [hunk 2]
```diff
@@ -567,12 +567,14 @@ void updateMemoryUsage()
...
         final PageWithPositionComparator pageWithPositionComparator;
 
+        boolean spillingWhenConvertingRevocableMemory;
         boolean resetPagesIndex;
         int pendingInputPosition;
 
         Optional<Page> currentSpillGroupRowPage;
         Optional<Spiller> spiller;
-        ListenableFuture<?> spillInProgress = immediateFuture(null);
+        // Spill can be trigger by Driver, by us or both. `spillInProgress` is not empty when spill was triggered but not `finishMemoryRevoke()` yet
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:576]: I assume "us" here means this window operator, especially `SpillablePagesToPagesIndexes#process`

**Pull request merged**